### PR TITLE
Extend token count to Embeddings and (non-streaming) BaseChatModel (#945)

### PR DESCRIPTION
**Pull request by benheckmann**
This PR addresses issue #945 that embeddings do not count toward the callback's token count.

I have added a callback handler to the base `Embeddings` class and implemented the calling of `on_llm_start` and `on_llm_end` for OpenAI.

I have also noticed and addressed that non-streaming chat completions using BaseChatModel also do not count towards the token count.

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
we probably want to do this:

1. on the base class
2. with separate methods (on_embedding_start)

this is probably a larger/more involved pr to be honest

**Comment by benheckmann** [comment]
First of all, thank you for taking a look and for your feedback!

Regarding 1. & 2.: Do you mean something like this would be the ideal solution:
```
class Embeddings(BaseModel, ABC):
    callback_manager: BaseCallbackManager = Field(default_factory=get_callback_manager)
    # ....

    @abstractmethod
    def _embed_documents(self, texts: List[str]) -> List[List[float]]:

    @abstractmethod
    def _embed_query(self, text: str) -> List[float]:

    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        self.callback_manager.on_embedding_start({"name": self.__class__.__name__}, texts)
        try:
            result = self._embed_documents(texts)
        except (Exception, KeyboardInterrupt) as e:
            self.callback_manager.on_embedding_error(e)
            raise e
        self.callback_manager.on_llm_end(result)

    def embed_query(self, text: str) -> List[float]:
        self.callback_manager.on_embedding_start({"name": self.__class__.__name__}, text)
        try:
            result = self._embed_documents(text)
        except (Exception, KeyboardInterrupt) as e:
            self.callback_manager.on_embedding_error(e)
            raise e
        self.callback_manager.on_llm_end(result)
```
(alternatively use a common wrapper for the two methods)


> this is probably a larger/more involved pr to be honest
Is this something that you would like to implement yourself? Or does it make sense for me to pursue this further?

**Comment by hwchase17** [comment]
> First of all, thank you so much for taking a look and for your feedback!
> 
> Regarding 1. & 2.: Do you mean something like this would be the ideal solution:
> 
> ```python
> class Embeddings(BaseModel, ABC):
>     callback_manager: BaseCallbackManager = Field(default_factory=get_callback_manager)
>     # ....
> 
>     @abstractmethod
>     def _embed_documents(self, texts: List[str]) -> List[List[float]]:
> 
>     @abstractmethod
>     def _embed_query(self, text: str) -> List[float]:
> 
>     def embed_documents(self, texts: List[str]) -> List[List[float]]:
>         self.callback_manager.on_embedding_start({"name": self.__class__.__name__}, texts)
>         try:
>             result = self._embed_documents(texts)
>         except (Exception, KeyboardInterrupt) as e:
>             self.callback_manager.on_embedding_error(e)
>             raise e
>         self.callback_manager.on_embedding_end(result)
> 
>     def embed_query(self, text: str) -> List[float]:
>         self.callback_manager.on_embedding_start({"name": self.__class__.__name__}, text)
>         try:
>             result = self._embed_documents(text)
>         except (Exception, KeyboardInterrupt) as e:
>             self.callback_manager.on_embedding_error(e)
>             raise e
>         self.callback_manager.on_embedding_end(result)
> ```
> 
> (alternatively use a common wrapper for the two methods...)
> 
> > this is probably a larger/more involved pr to be honest
> 
> Is this something that you would like to implement yourself? Or does it make sense for me to pursue this further?

yeah i think something you described is ideal. would also make it easy to put caching in the base class. in fact theres another pr open to do just that, which also puts this wrapper on the base class (https://github.com/hwchase17/langchain/pull/1930)

if you want to pursue it further, that would work great! i would suggest:

1. a pr to create this base class wrapper, even if it just does only the below to start:
```
def embed_documents(self, texts: List[str]) -> List[List[float]]:
        return self._embed_documents(texts)
```

(this will make it easy to do caching at the same time)

2. a separate method to add those endpoints and hook them up (i can help with this if needed, theres a bit of plumbing to be done, but if you want to take stab i dont think should be too bad (famous last words)

cc @agola11 as well, as he did a lot of the callback stuff

**Review by agola11 — COMMENTED** [review]
Adding the embeddings callbacks sounds fine to me, although we should add them as non-abstract methods to avoid breaking changes. They can simply be empty methods in the `BaseCallbackHandler` class

File: langchain/callbacks/openai_info.py [hunk 1]
```diff
@@ -32,12 +32,9 @@ def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
         if response.llm_output is not None:
             if "token_usage" in response.llm_output:
                 token_usage = response.llm_output["token_usage"]
-                if "total_tokens" in token_usage:
-                    self.total_tokens += token_usage["total_tokens"]
-                if "prompt_tokens" in token_usage:
-                    self.prompt_tokens += token_usage["prompt_tokens"]
-                if "completion_tokens" in token_usage:
-                    self.completion_tokens += token_usage["completion_tokens"]
+                self.total_tokens += token_usage.get("total_tokens", 0)
```
[the remark this answers is not in the document]
Comment (benheckmann) [reply langchain/callbacks/openai_info.py:35]: Yes, just to save a few lines of code.

**Comment by benheckmann** [comment]
Hi and sorry for the delay!

I have implemented the wrapper as discussed for the `Embeddings` (base) class. None of the new methods are abstract so there should be no breaking changes.

As discussed, this should as a starting point for implementing the token count as well as the caching (#1913) for embeddings.

**Comment by benheckmann** [comment]
Would be great if either of you could take a look @hwchase17  or @agola11.

**Comment by benheckmann** [comment]
Closing this now.

**Pull request closed without merging**